### PR TITLE
Add Explicit Workload Deletion

### PR DESCRIPTION
**Pull request by razo7**
Explicit/Manuel workload deletion.

Deleting the pods and their volume attachments in order to signal the scheduler that this node's workloads should be rescheduled on a different node.
Includes e2e and unit tests which adds volume attachments and pod to the unhealthy node in order to verify if they have deleted as expected.

[ECOPROJECT-1473](https://issues.redhat.com//browse/ECOPROJECT-1473)

### CODE REVIEW

File: controllers/fenceagentsremediation_controller.go
Comment (clobrano) [inline controllers/fenceagentsremediation_controller.go:164]: typo s/Manuel/Manual

(also few lines below)

File: pkg/utils/resources.go
Comment (clobrano) [inline pkg/utils/resources.go:16]: typo s/loggerResouces/loggerResources/

**Review by clobrano — COMMENTED** [review]
only some typos from my point of view
/lgtm 
Giving others a chance to review as well, feel free to unhold
/hold

File: pkg/utils/resources.go
Comment (slintes) [inline pkg/utils/resources.go:20]: The reboot is the fencing.
This is just improving rescheduling time.

File: pkg/utils/resources.go
Comment (slintes) [reply pkg/utils/resources.go:16]: I would call it just `logger` or even `log`. Why do we need the package name in the name?

**Review by slintes — CHANGES_REQUESTED** [review]

File: controllers/fenceagentsremediation_controller.go
Comment (mshitrit) [reply controllers/fenceagentsremediation_controller.go:163]: IIUC that's in line 150 `outputRes, outputErr, err := r.Executor.Execute(pod, cmd)`

Comment (slintes) [reply controllers/fenceagentsremediation_controller.go:163]: - a *non graceful* shutdown might have happened already, depending on node issue which made it unhealthy
- we *deal with* the *non graceful* shutdown by resource deletion (or out of service taint)
- the reboot command is more likely a graceful shutdown (https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown)
- last but not least: the reboot command was executed before this comment already, and we don't do any other shutdown after this comment 🤷🏼‍♂️

File: pkg/utils/resources.go
Comment (razo7) [reply pkg/utils/resources.go:20]: - This is just improving rescheduling time.
What about stateful pods (pods with PVC)? IIUC these kind of pods won't move/evict to another node

File: pkg/utils/resources.go
Comment (razo7) [reply pkg/utils/resources.go:74]: SNR rights are still preserved somehow with [line](https://github.com/medik8s/fence-agents-remediation/pull/63/files#diff-58c65d0ac78159b5a891995b8185dda69a88d7259470696f9de752dc52f7257fR3).
Next release we can move this logic, and more (such as taint logic from last PRs) to the common repo.

File: controllers/fenceagentsremediation_controller.go
Comment (razo7) [reply controllers/fenceagentsremediation_controller.go:163]: I believe this comment can be improved and reflect on the pods removal rather than Node Shutdown

File: pkg/utils/resources.go
Comment (razo7) [reply pkg/utils/resources.go:16]: > Why do we need the package name in the name?

Just for easier identification of the log's source.

Comment (slintes) [reply pkg/utils/resources.go]: I mean the var name :)

**Review by slintes — COMMENTED** [review]
looks good now, what about tests? :)

File: pkg/utils/resources.go
Comment (slintes) [inline pkg/utils/resources.go:19]: we need to pass the context from the reconciler as parameter here, and use it in all client calls

**Review by slintes — CHANGES_REQUESTED** [review]
something went wrong while rebasing, at least one commit is missing!

File: pkg/utils/resources.go
Comment (slintes) [reply pkg/utils/resources.go:45]: at least one commit is lost, it was addressing this comment ^^
https://github.com/medik8s/fence-agents-remediation/commit/b1574e8321863d77235a40ced5cdad766ed0807b

Comment (razo7) [reply pkg/utils/resources.go:45]: > I understand the ListOptions docs like an empty namespace means all namespaces.

You are right but my old comment that addressed your suggestion introduced an error on pod deletion -> `the server does not allow this method on the requested resource`.
When the namespace of ListOptions is [empty](https://github.com/medik8s/fence-agents-remediation/pull/63/files#diff-58c65d0ac78159b5a891995b8185dda69a88d7259470696f9de752dc52f7257fR26), and we delete as follows:

```suggestion
if err := r.DeleteAllOf(ctx, pod, deleteOptions); err != nil {
		log.Error(err, "failed to delete pods of unhealthy node") //, "namespace", ns.Name)
		return err
	}
```

Comment (slintes) [reply pkg/utils/resources.go:45]: oh, I see...

https://github.com/kubernetes-sigs/controller-runtime/issues/1842

File: controllers/fenceagentsremediation_controller_test.go
Comment (clobrano) [inline controllers/fenceagentsremediation_controller_test.go:172]: The intention could be clearer if writing something like
```
nodeKey.Name = underTestFAR.Name
```
?

File: controllers/fenceagentsremediation_controller_test.go
Comment (clobrano) [inline controllers/fenceagentsremediation_controller_test.go:216]: typo: *should reside or *should be

File: controllers/fenceagentsremediation_controller_test.go
Comment (clobrano) [inline controllers/fenceagentsremediation_controller_test.go:228]: typo: *its

also below and in `far_e2e_test.go`

File: test/e2e/far_e2e_test.go
Comment (clobrano) [inline test/e2e/far_e2e_test.go:34]: typo: s/succeesRebootMessage/successRebootMessage

File: test/e2e/far_e2e_test.go
Comment (slintes) [inline test/e2e/far_e2e_test.go:252]: nit: why don't you set this above?

File: controllers/fenceagentsremediation_controller_test.go
Comment (clobrano) [inline controllers/fenceagentsremediation_controller_test.go:258]: I understand this function verifies if the resources have been already deleted, but its actual job is to clean them up, and the name should reflect that.
What something simple like `deleteResources`?

The same goes for the other similar functions here and in e2e test

**Review by slintes — APPROVED** [review]
one nit, I leave it up to you if you want to address it

/hold

**Review by clobrano — CHANGES_REQUESTED** [review]

File: controllers/fenceagentsremediation_controller_test.go
Comment (razo7) [reply controllers/fenceagentsremediation_controller_test.go:228]: I used it's when I wanted to use its... I will add that (and other typo comments) in a follow up PR if you don't mind

**Review by clobrano — COMMENTED** [review]
/lgtm

**Review by clobrano — APPROVED** [review]

**Pull request merged**
